src/main/java/dev/corestone/mapprotect/regions/regionmanagers/player_managers/ExplosionDamageHandler.java

```diff
@@ -63,17 +66,28 @@ public void explosionDamageListenerBlock(BlockExplodeEvent event) {
     }
 
     @EventHandler
-    public void explosionDamageListener(EntityDamageEvent event){
+    public void explosionDamageByEntityListener(EntityDamageByEntityEvent event){
         if (canExplode)return;
         if (region.getState() == RegionState.IDLE)return;
-        if (event.getCause() == EntityDamageEvent.DamageCause.BLOCK_EXPLOSION || event.getCause() == EntityDamageEvent.DamageCause.ENTITY_EXPLOSION) {
+        if (event.getCause() == EntityDamageEvent.DamageCause.ENTITY_EXPLOSION && region.getBox().contains(event.getDamager().getLocation().toVector())){
+            event.setCancelled(true);
+            if(event.getEntity() instanceof Player){
+                event.getEntity().sendMessage(Colorize.format(explosionDamageDenyMessage));
+            }
+        }
+    }
+    //event.getCause() == EntityDamageEvent.DamageCause.BLOCK_EXPLOSION ||
+    @EventHandler
+    public void explosionDamageByBlockListener(EntityDamageByBlockEvent event){
+        if (canExplode)return;
+        if (region.getState() == RegionState.IDLE)return;
+        if (event.getCause() == EntityDamageEvent.DamageCause.ENTITY_EXPLOSION && region.getBox().contains(event.getDamager().getLocation().toVector())){
```

Comment:
I think there is a difference between a damager and the damaged (e.g. event.getDamager() vs event.getEntity()). The entity is the entity getting damaged. Here, we mostly care about entities inside of the box. So in short, change this to event.getEntity() instead of event.getDamager(). Event.getDamager() is the block.